File: hto/12/hto_12.cpp
#include <iostream>
#include <string>
#include <vector>
#include <climits>
#include <cstring>
using namespace std;

class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        int rows = board.size();
        int cols = board[0].size();
        vector<vector<bool>> visited(rows, vector<bool>(cols, false));
        for (int i = 0; i < board.size(); ++i) {
            for (int j = 0; j < board[0].size(); ++j) {
                if (dfs(board, word, visited, i, j, 0)) return true;
            }
        }
        return false;
    }

private:
    bool dfs(vector<vector<char>> &board, string &word,
             vector<vector<bool>> &visited, int i, int j, int k) {
        if (k == word.size()) return true;
        if (i < 0 || i == board.size() || j < 0 || j == board[0].size() ||
            visited[i][j] || board[i][j] != word[k]) return false;
        visited[i][j] = true;
        bool res = dfs(board, word, visited, i + 1, j, k + 1) ||
                   dfs(board, word, visited, i - 1, j, k + 1) ||
                   dfs(board, word, visited, i, j + 1, k + 1) ||
                   dfs(board, word, visited, i, j - 1, k + 1);
        visited[i][j] = false;
        return res;
    }
};

int main() {
    vector<vector<char>> board = {{'A', 'B', 'C', 'E'},
                                  {'S', 'F', 'C', 'S'},
                                  {'A', 'D', 'E', 'E'}};
    string word = "ABCCED";
    Solution s;
    cout << boolalpha << s.exist(board, word) << endl;
}

File: hto/12/hto_12.go
func exist(board [][]byte, word string) bool {
    visited := make([][]bool, len(board))
    for i := range visited {
        visited[i] = make([]bool, len(board[0]))
    }
    for i := 0; i < len(board); i++ {
        for j := 0; j < len(board[0]); j++ {
            if dfs(board, word, visited, i, j, 0) {
                return true
            }
        }
    }
    return false
}

func dfs(board [][]byte, word string, visited [][]bool, i, j, k int) bool {
    if i < 0 || i == len(board) || j < 0 || j == len(board[0]) || board[i][j] != word[k] || visited[i][j] {
        return false
    }
    if k == len(word) - 1 {
        return true
    }
    visited[i][j] = true
    res := dfs(board, word, visited, i + 1, j, k + 1) || 
           dfs(board, word, visited, i - 1, j, k + 1) || 
           dfs(board, word, visited, i, j + 1, k + 1) || 
           dfs(board, word, visited, i, j - 1, k + 1)
    visited[i][j] = false
    return res
}
